#! /bin/sh /usr/share/dpatch/dpatch-run
## dosfix-ldapbackend.dpatch by Matthijs Mohlmann <matthijs@cacholong.nl>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: DoS fix in the ldapbackend.

@DPATCH@
diff -urNad --exclude=CVS --exclude=.svn ./modules/ldapbackend/ldapbackend.cc /tmp/dpep-work.cNeNy7/pdns-2.9.17/modules/ldapbackend/ldapbackend.cc
--- ./modules/ldapbackend/ldapbackend.cc	2005-07-07 10:03:10.088280864 +0200
+++ /tmp/dpep-work.cNeNy7/pdns-2.9.17/modules/ldapbackend/ldapbackend.cc	2005-07-07 10:12:06.673707480 +0200
@@ -125,11 +125,14 @@
 {
 	string dn;
 	string filter;
+  string qesc;
+
+  qesc = toLower(m_pldap->escape( target));
 
 
 	// search for SOARecord of target
 	dn = getArg( "basedn" );
-	filter = "(associatedDomain=" + target + ")";
+	filter = "(associatedDomain=" + qesc + ")";
 	m_msgid = m_pldap->search( dn, LDAP_SCOPE_SUBTREE, filter, (const char**) ldap_attrany );
 	m_pldap->getSearchEntry( m_msgid, m_result, true );
 
@@ -140,7 +143,7 @@
 	}
 
 	prepare();
-	filter = "(associatedDomain=*." + target + ")";
+	filter = "(associatedDomain=*." + qesc + ")";
 	DLOG( L << Logger::Debug << m_myname << " Search = basedn: " << dn << ", filter: " << filter << endl );
 	m_msgid = m_pldap->search( dn, LDAP_SCOPE_SUBTREE, filter, (const char**) ldap_attrany );
 
@@ -270,7 +273,7 @@
 	vector<string> parts;
 
 
-	qesc = toLower( qname );
+	qesc = toLower( m_pldap->escape(qname) );
 	filter = "(associatedDomain=" + qesc + ")";
 
 	if( qtype.getCode() != QType::ANY )
